alternator/executor.cc

```diff
@@ -2354,23 +2355,39 @@ mutation put_or_delete_item::build(schema_ptr schema, api::timestamp_type ts) co
             row.cells().apply(*cdef, atomic_cell::make_live(*cdef->type, ts, std::move(c.value)));
         }
     }
+    auto attrs = attrs_column(*schema);
     if (!attrs_collector.empty()) {
         auto serialized_map = attrs_collector.to_mut().serialize(*attrs_type());
-        row.cells().apply(attrs_column(*schema), std::move(serialized_map));
+        row.cells().apply(attrs, std::move(serialized_map));
     }
     // To allow creation of an item with no attributes, we need a row marker.
     row.apply(row_marker(ts));
     // PutItem is supposed to completely replace the old item, so we need to
-    // also have a tombstone removing old cells. We can't use the timestamp
-    // ts, because when data and tombstone tie on timestamp, the tombstone
-    // wins. So we need to use ts-1. Note that we use this trick also in
-    // Scylla proper, to implement the operation to replace an entire
-    // collection ("UPDATE .. SET x = ..") - see
-    // cql3::update_parameters::make_tombstone_just_before().
-    if (use_partition_tombstone) {
-        m.partition().apply(tombstone(ts-1, gc_clock::now()));
-    } else {
-        row.apply(tombstone(ts-1, gc_clock::now()));
+    // also have a tombstone removing old cells. Important points:
+    // 1) Alternator's schema is dynamic, therefore we store data in a map
+    //    in column :attrs. So, Alternator tables only have columns pk, ck, and
+    //    :attrs. Since we're replacing a row, invalidating only :attrs is
+    //    enough. Alternator base tables also had columns for LSI keys and GSI
+    //    keys. New tables no longer have such columns, but old tables created
+    //    in the past may still have them.
+    // 2) We use a collection tombstone for the :attrs column instead of a row
+    //    tombstone. While a row tombstone would also replace the data, it has
+    //    an undesirable side effect for CDC, which would report it as a 
+    //    separate deletion event. To model PutItem's "replace" semantic, we
+    //    leverage a corner case: a collection tombstone at ts-1 paired with an
```

Comment:
Note that you removed from the original comment the explanation why **ts-1** is the timestamp we want to use (and that it's not an Alternator invention but used already in CQL when replacing a collectionn).